Add GEP-917 to document conformance principles (#918)
kubernetes-sigs/gateway-api: site-src/geps/gep-917.md

similar efforts in the future. Ensuring that whatever we build is reusable for
other projects will help to lift everyone who works with CRDs.

### What do we need out of conformance for Gateway API?
Must have:

- Support for only implementing some of the Gateway API CRD objects. Some objects,
like Gateway itself, are required for all implementations, but implementations
may choose what Route objects they support. There will probably be some common
sets of Route objects supported across similar implementations, but this proposal
expects that we will make calls about what to call that common experience at a
later date.
- Support for fields or features that have Core, Extended, and
ImplementationSpecific support. In particular, it must be possible for
implementations to only support some subset of Extended fields, and to be able
to use the framework for their own ImplementationSpecific features if required.
- A testing suite that can validate that an implementation meets the conformance
profiles it claims.
- A way to retrieve conformance information in a machine-parseable way.

## Proposal

### Conformance profiles principles

#### Basics

> The Gatway API project defines conformance purely in terms of what objects an

Reviewer: ```suggestion
The Gateway API project defines conformance purely in terms of what objects an
```

implementation supports.

> By "supports" in this context, we mean the functionality covered by Core support.
> Extended support fields are handled by a separate mechanism.

Reviewer: ```suggestion
To support a resource, an implementation must support all "Core" functionality defined by the resource. Support for "Extended" functionality will be indicated separately.
```

All implementations must support all the Core functions on the following objects:

> - GatewayClass

Reviewer: Tiny nit: Need new line before this for markdown list parsing

- Gateway
- ReferencePolicy

The following objects are optional to support, but have defined behavior if you
do:

> - HTTPRoute

Reviewer: Tiny nit: Need new line before this for markdown list parsing

- TLSRoute
- TCPRoute
- UDPRoute

> For all of these objects, we should aim to have the usual range of tests for

Reviewer 2: ```suggestion
For all of these resources, we should aim to have the usual range of tests for
```
Use a replace-all for this.

both the happy and unhappy paths for various types of operations.

The conformance is versioned - it tracks the required features for a specific
version of the API, and must be included in and updated by a version bump in the
bundle version of the Gateway API. (The _bundle version_ is the thing that we
mark as a "release", that looks like `v0.4.0`, not `v1alpha2`). 

> This will enable an implementation to say that it supports version whatever of

Reviewer: ```suggestion
This will enable an implementation to say that it supports a specific version of
```

the Gateway API. This is again similar to upstream in that implementations need
to submit conformance test results for each version of Kubernetes they support.

> Because the support is defined in terms of the objects that an implementation

Reviewer: ```suggestion
Because the support is defined in terms of the resources that an implementation
```

supports, the conformance is composable, and orthogonal for each object type.
It's valid to only support TLSRoute and not HTTPRoute, for example, although

> it would be odd.

Reviewer: Tiny nit: Maybe focus on a more common example such as supporting HTTPRoute but not TCPRoute.

Reviewer 2: Also, remove the "it would be odd", we may be potentially shooting down some cases. Others in the community will be (are?) more creative and use this API in interesting ways.

#### Interaction with existing support levels

Conformance definitions will ensure that an implementation can provide all the
features currently marked as "Core" support in the API documentation.

Fields marked "Extended" support will eventually have conformance tests that
lock in the behavior of that feature, and there will be a mechanism for implementations
to tell the testing framework what extended fields they support.

#### Testing framework

The Gateway API project will provide a set of tests and harness to run them, such

> that an implementation may point the test harness at a Gateway or set of Gateways

Reviewer: I've actually thought of conformance targeting a GatewayClass and then creating/testing Gateways, would that be workable?

Reviewer 2: Maybe?
In-cluster proxies will need something akin to an operator for that to work.
Kong is not considering Gateway lifecycle management in the first iteration.

Reviewer: If Kong is deployed behind a Service, will it be sufficiently aware of that to set an IP in Gateway status? I believe that's how Istio's currently setting this status, but not sure, @howardjohn would know best.

Author: I think that just putting in here that the testing may be pointed at a GatewayClass or Gateway will cover us no matter what.

But I think that full dynamic Gateway creation may be a ways away for a few implementations at least, Kong and Contour among them. So we should have a way to set whether a new Gateway can be created or not as well.

> managed by that implenentation and have the testing framework deliver a report

Reviewer: ```suggestion
managed by that implementation and have the testing framework deliver a report
```

(As per our upstream KEP). It's acceptable to begin with a small set and expand
outward while the project is in beta, but having a full set of conformance tests
that cover most of the API scope should be a requirement for declaring the API
stable.

#### Certification process

The Gateway API project will provide a process by which an implementation may
submit the results of a run of the conformance test suite to a centralized,
open repository, and once verified, these results will be used to maintain a
canonical list of certified conformant implementations.

Ideally, this process should be handled using similar methods to upstream
Kubernetes, while also learning what we can from what the upstream conformance
efforts wish they could improve.

### Initial conformance profiles

These initial conformance profiles are a rough sketch, and are expected to change
as we build out the actual processes around them. This is intended as a starting
point for discussion, no more.

In particular, the names are placeholders, and suggestions for change are very
welcome.

#### Layer 7 conformance profile

Must support the following objects:
- GatewayClass
- Gateway
- ReferencePolicy
- HTTPRoute

> - TLSRoute

Author: Once we have conformance tests, it should be, but in practice right now probably isn't.

similar efforts in the future. Ensuring that whatever we build is reusable for
other projects will help to lift everyone who works with CRDs.

### What do we need out of conformance for Gateway API?
Must have:

- Support for different types of conformance, at a minimum Layer 4 and Layer 7
Gateway API support. We call this “Conformance Profiles”, which is
inline with upstream Kubernetes.
- Support for fields or features that have Core, Extended, and
ImplementationSpecific support. In particular, it must be possible for
implementations to only support some subset of Extended fields, and to be able
to use the framework for their own ImplementationSpecific features if required.
- A testing suite that can validate that an implementation meets the conformance
profiles it claims.
- A way to retrieve conformance information in a machine-parseable way.

## Proposal

### Conformance profiles principles

#### Basics
We have two conformance profiles (to begin with), covering what we will loosely
call "Layer 7" (meaning focussed around HTTP and HTTPS, like a traditional
Ingress controller), and "Layer 4", focussed around forwarding TCP and UDP

> traffic.

Author: The idea of profiles is to make it easier to understand what Routes can move between Gateways. If the Gateways have the same profile, if you are using features covered by conformance (ie Core or Extended), then you can safely move those Routes between those Gateways. 

I think that doing it the way you're talking about will mean that users will need to take a "guess I'll try it out and see if it works" approach, which I don't think is desirable.

Obviously, we can't mandate that everything works the same exactly between different controllers, but the support levels and conformance profiles should mean that application developer persona users can be confident that a Route can safely move between Gateways, regardless of implementation.

To put this another way, in my mind, one of the primary goals of the Gateway API project is to work towards Route portability between Gateways. The API is complex enough that we can't guarantee this across the board, but by providing subsets of common functionality, I think we can come pretty close. That's what the profiles are intended to achieve.

> traffic.

Author: We had a great discussion about this at the community meeting today, and I agreed that I'll rework this document to put the focus on the per-object conformance, and let the conformance profiles happen organically later. I'd be pretty shocked if we don't end up with some way to talk about the common sets of objects supported later, but this was probably a premature optimization on my part.
